packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,413 @@
+// Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import assert from 'assert';
+import fs from 'fs';
+import path from 'path';
+import * as zlib from 'zlib';
+import { JsonRpcSuccessResponse } from '@cosmjs/json-rpc';
+import { Registry } from '@cosmjs/proto-signing';
+import { Log } from '@cosmjs/stargate/build/logs';
+import { adaptor37 } from '@cosmjs/tendermint-rpc/build/tendermint37/adaptor';
+import {
+  BlockResponse,
+  BlockResultsResponse,
+} from '@cosmjs/tendermint-rpc/build/tendermint37/responses';
+import KyveSDK, { KyveLCDClientType } from '@kyvejs/sdk';
+import { SupportedChains } from '@kyvejs/sdk/src/constants'; // Currently these types are not exported
+import {
+  PoolResponse,
+  QueryPoolsResponse,
+} from '@kyvejs/types/lcd/kyve/query/v1beta1/pools';
+import { delay, getLogger } from '@subql/node-core';
+import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
+import { BlockContent } from '../../indexer/types';
+import { LazyBlockContent } from '../cosmos';
+import { BundleDetails } from './kyveTypes';
+
+const BUNDLE_TIMEOUT = 10000; //ms
+const POLL_TIMER = 3; // sec
+const MAX_COMPRESSION_BYTE_SIZE = 2 * 10 ** 9;
+
+const parseDecimal = (value: string) => parseInt(value, 10);
+
+const logger = getLogger('kyve');
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const { version: packageVersion } = require('../../../package.json');
+
+interface UnZippedKyveBlockReponse {
+  value: { block: any; block_results: any };
+  key: string;
+}
+
+export class KyveApi {
+  private readonly lcdClient: KyveLCDClientType;
+  private respAdaptor = adaptor37.responses;
+  private poolId: string;
+  private currentBundleId = -1;
+  private cachedBundleDetails: BundleDetails;
+
+  private constructor(
+    private chainId: string,
+    private endpoint: string,
+    private readonly storageUrl: string,
+    private readonly kyveChainId: SupportedChains,
+    private readonly tmpCacheDir: string,
+  ) {
+    this.lcdClient = new KyveSDK(this.kyveChainId, {
+      rpc: this.endpoint,
+    }).createLCDClient();
+    this.storageUrl = storageUrl;
+  }
+
+  static async create(
+    chainId: string, // chainId for indexing chain
+    endpoint: string,
+    storageUrl: string,
+    kyveChainId: SupportedChains,
+    tmpCacheDir: string,
+  ): Promise<KyveApi> {
+    const kyve = new KyveApi(
+      chainId,
+      endpoint,
+      storageUrl,
+      kyveChainId,
+      tmpCacheDir,
+    );
+    await kyve.setPoolId();
```

Comment:
The way this is setup means that it could be called multiple times. Instead passing the result of this in as a constructor argument would be a better way.